### DISCUSSION TITLE
Thread Router Multi-pan Communication

### DISCUSSION THREAD

**Discussion by orangenlimer159** [openthread/openthread#9770] [Q&A] [answered]
Hello all,

I'm pretty new to OpenThread and would like to seek some guidance and clarification on this one:

I'm currently working with several Minimal Thread Devices (MTDs) and a Thread router. Is it possible to configure the router to support multiple PAN IDs simultaneously? If so, is there any relevant documentation or examples that I can refer to?

Thanks in advance!

**Comment by jwhui** [comment, marked as answer]
To support multiple Thread networks on a single Thread router, ideally the Thread router will have multiple Thread radios, one to support each Thread network. Typically, a Thread radio can only be listening on a single channel, so having multiple Thread radios allows each Thread network to operate independently, including on different channels.

**Reply by Rutvik2598** [reply]
How can I make one radio to launch multiple instances? I see some flags like OT_MULTIPAN_RCP=1 OT_MULTIPLE_INSTANCE=1, but I am not able to figure out how I can configure this. I tried building the ot-rcp binary using the following command

**./script/build nrf52840 USB_trans -DOT_BOOTLOADER=USB -DOT_FTD=OFF -DOT_MTD=OFF -DOT_RCP=ON -DOT_MULTIPAN_RCP=1 -DOT_MULTIPLE_INSTANCE=1**

I converted the binary to hex and converted it to uf2 format to flash the radio dongle. But somehow after flashing my nrf52840, it it not getting recognized as a USB in my raspberry pi 3. 

Could you please guide me, how I can configure my nrf52840 dongle to support multiple instances?

Also is there a way that I can also simulate this?

Thanks in advance for the help. :)

**Comment by Raj-mac24** [comment]
Hi,  Were you able to successfully establish the network where multiple PAN-IDs are handled on a single router in openthread?

**Reply by jwhui** [reply]
Are you looking to support multiple Thread networks with a single 802.15.4 radio? We generally recommend using multiple 802.15.4 radios, especially in cases where the different networks are operating on different channels.

**Comment by Raj-mac24** [comment]
Do you think, I mean is it possible to configure multiple pan-ids on a
single router? If I change the open thread code?

**Reply by jwhui** [reply]
Each Thread network has its own PAN ID. Multiple Thread networks can be done by having multiple OpenThread instances.

**Comment by Raj-mac24** [comment]
But can I configure multiple pan-ids(ESP32) on a single router (single open
thread network) ?

**Reply by jwhui** [reply]
As mentioned in [above](https://github.com/orgs/openthread/discussions/9770#discussioncomment-12281130), each Thread network has its own PAN ID. It is possible for a single device to support multiple Thread interfaces, just like supporting multiple Ethernet, Wi-Fi, or Cellular interface.

Also, please respond to the original thread, rather than creating new suggested answers.

**Reply by chshu** [reply]
@orangenlimer159 @Raj-mac24 The multipan feature on ESP32 802.15.4 SoCs allows a device to operate in multiple 15.4 PANs simultaneously. However, all PANs must be on the same channel.

A single OpenThread network has only one PAN ID, but it is possible to support multiple Thread networks using multiple OpenThread instances.

For assistance with your specific requirements, please contact us via [Contact Us](https://www.espressif.com/en/contact-us/sales-questions).

**Reply by anhns26693** [reply]
Hi @chshu 
Do you have any guideline for "multiple Thread networks using multiple OpenThread instances." with multiple Thread radio device ?
Thanks so much !!!

**Reply by Rutvik2598** [reply]
@anhns26693 You can run multiple OpenThread instances on the same device (for example, a Raspberry Pi), with each instance bound to a different radio device. The setup process is essentially the same as launching a single instance—just repeat it for each radio interface.

**Reply by anhns26693** [reply]
Hi @Rutvik2598 

my case is running two instance of otbr-agent as services on same raspi with seperate thread radio usb

Device : Raspberry PI 4B
OS : Debian 11 bulleys
Matter source : connectedhomeip v1.4.0.0
Some matter Thread device build for testing with Thread network
ot-br-posix: https://github.com/openthread/ot-br-posix.git tag/thread-reference-20230710
    sudo INFRA_IF_NAME=wlan0 OTBR_OPTIONS="-DOT_THREAD_VERSION=1.3" ./script/setup
Thread radio : nRF52840 usb dongle (2 devices)

I run two otbr-agent as service

$ cat /etc/default/otbr-agent
OTBR_AGENT_OPTS="-I wpan0 -B wlan0 spinel+hdlc+uart:///dev/ttyACM0 trel://wlan0"
OTBR_NO_AUTO_ATTACH=0


$ cat /etc/default/otbr-agent1
OTBR_AGENT_OPTS1="-I wpan1 -B wlan0 spinel+hdlc+uart:///dev/ttyACM1 trel://wlan0"
OTBR_NO_AUTO_ATTACH=0

1. I start and run otbr-agent first
2. I start and run otbr-agent1 then
3. Turn on two console for seeing log by command journalctl -u otbr-agent -f and journalctl -u otbr-agent1 -f
4. ot-ctl -I wpan0 state --> work normally
5. ot-ctl -I wpan0 ifconfig up --> work normally
6. ot-ctl -I wpan1 state --> ( on console log of  otbr-agent1 ) matter otbr-agent[49708]: 00:01:07.269 [W] Platform------: Daemon read: Connection reset by peer
7. ot-ctl -I wpan1 ifconfig up --> ( on console log of  otbr-agent1 ) matter otbr-agent[49708]: 00:01:17.807 [C] Platform------: Failed to bind UDP socket: Address already in use.

Do you have any advise for me to resolve this issue ?
Thank you so much.

**Reply by jwhui** [reply]
See https://github.com/orgs/openthread/discussions/11947#discussioncomment-14435482

**Comment by piotrasd** [comment]
A Step-by-Step Guide to Multipan RCP on ESP-IDF 6.x

> **Status**: Working on hardware. Zigbee coordinator + Thread Border Router sharing one ESP32-H2 802.15.4 radio via Spinel IID multiplexing.
>
> **Related**: [espressif/esp-idf#13815](https://github.com/espressif/esp-idf/issues/13815) — Espressif confirmed this is possible but never published implementation details. This guide fills that gap.

---

## Table of Contents

1. [Architecture Overview](#1-architecture-overview)
2. [How Spinel IID Multiplexing Works](#2-how-spinel-iid-multiplexing-works)
3. [H2 RCP Firmware: Multi-Instance NCP](#3-h2-rcp-firmware-multi-instance-ncp)
4. [S3 Host: Spinel Mux Layer](#4-s3-host-spinel-mux-layer)
5. [S3 Host: MuxSpinelInterface](#5-s3-host-muxspinelinterface)
6. [ESP-IDF Platform Patches Required](#6-esp-idf-platform-patches-required)
7. [Build Configuration](#7-build-configuration)
8. [Debugging Guide](#8-debugging-guide)
9. [Known Limitations](#9-known-limitations)
10. [How We Got Here: The Investigation](#10-how-we-got-here-the-investigation)

---

## 1. Architecture Overview

```
ESP32-S3 (WiFi host)                     ESP32-H2 (802.15.4 RCP)
+-----------------------+                +------------------------+
| Zigbee Coordinator    |                | otInstanceInitMultiple(0) |
|   esp-zigbee-lib      |--IID 0--+     |   (Zigbee radio)       |
|   RadioSpinel         |         |     +------------------------+
+-----------------------+         |                              
                                  |     UART (460800 baud)       
+-----------------------+    +----+----+  HDLC framing           
| Thread Border Router  |    | spinel  |                         
|   OpenThread FTD      |--IID 1--| _mux    |--+------------------------+
|   RadioSpinel         |    +---------+  | otNcpHdlcInitMulti()   |
+-----------------------+         |     | NCP dispatches by IID  |
                                  |     +------------------------+
+-----------------------+         |     | otInstanceInitMultiple(1) |
| MuxSpinelInterface    |---------+     |   (Thread radio)       |
|   eventfd + queue     |               +------------------------+
+-----------------------+
```

**Key principle**: The H2 runs a standard OpenThread multipan NCP with 2 instances. The S3 host multiplexes Spinel frames by IID (Interface ID) over a shared UART. Each protocol stack on the S3 thinks it has its own dedicated radio.

## 2. How Spinel IID Multiplexing Works

The Spinel protocol header byte encodes an Interface ID (IID) in bits 5-4:

```
Spinel Header Byte:
  Bit 7-6: Flag (always 0b10)
  Bit 5-4: IID (0-3)
  Bit 3-0: TID (transaction ID, 0-15)
```

| IID | Assignment | Purpose |
|-----|-----------|---------|
| 0 | Zigbee | Default active radio instance |
| 1 | Thread | Secondary radio instance |
| 2 | (unused) | Available for future use |
| 3 | Broadcast | H2 sends RESET/status on IID 3 to all consumers |

When the H2 NCP receives a Spinel frame, it extracts the IID from the header, maps it to the correct `otInstance`, and processes the command in that instance's context. This is standard OpenThread behavior (`ncp_base.cpp:458: mInstance = IidToInstance(mCurCommandIid)`).

## 3. H2 RCP Firmware: Multi-Instance NCP

The H2 runs a custom RCP firmware that initializes two OpenThread instances:

```c
// h2_rcp/main/h2_rcp_main.c
#define H2_RCP_INSTANCE_COUNT 2

static otInstance *s_instances[H2_RCP_INSTANCE_COUNT];

void app_main(void) {
    // ... platform init ...
    
    for (uint8_t i = 0; i < H2_RCP_INSTANCE_COUNT; ++i) {
        s_instances[i] = otInstanceInitMultiple(i);
    }
    
    otNcpHdlcInitMulti(s_instances, H2_RCP_INSTANCE_COUNT, ncp_uart_send);
    
    // Store primary instance for IDF compatibility
    esp_openthread_set_instance(s_instances[0]);
    
    h2_rcp_launch_mainloop(s_instances[0]);
}
```

### Critical: Mainloop Must Process ALL Instances

The mainloop must check tasklets and process alarms for every instance, not just the primary:

```c
// WRONG: only processes instance 0
while (otTaskletsArePending(instance)) {
    otTaskletsProcess(instance);
}

// CORRECT: process all instances
for (uint8_t i = 0; i < H2_RCP_INSTANCE_COUNT; ++i) {
    while (otTaskletsArePending(s_instances[i])) {
        otTaskletsProcess(s_instances[i]);
    }
}
```

Without this, instance 1's SubMac CSMA backoff timer never fires and radio TX silently fails.

### Multipan Platform Callbacks

The H2 firmware provides these callbacks for the NCP:

```c
otError otPlatMultipanGetActiveInstance(otInstance **aInstance) {
    *aInstance = s_active_instance;
    return OT_ERROR_NONE;
}

otError otPlatMultipanSetActiveInstance(otInstance *aInstance, bool aCompletePending) {
    if (s_active_instance == aInstance) return OT_ERROR_ALREADY;
    s_active_instance = aInstance;
    otPlatMultipanSwitchoverDone(aInstance, true);  // instant on single radio
    return OT_ERROR_NONE;
}
```

`otPlatMultipanIidToInstance` and `otPlatMultipanInstanceToIid` are provided by OpenThread's `multipan_platform.cpp`.

### H2 Build Configuration

Key sdkconfig options for the H2 RCP:
```
CONFIG_OPENTHREAD_RADIO=y
CONFIG_OPENTHREAD_CONFIG_MULTIPLE_INSTANCE_ENABLE=1
CONFIG_OPENTHREAD_CONFIG_MULTIPLE_STATIC_INSTANCE_ENABLE=1
CONFIG_OPENTHREAD_CONFIG_MULTIPAN_RCP_ENABLE=1
```

## 4. S3 Host: Spinel Mux Layer

The S3 runs a Spinel IID multiplexer (`spinel_mux.c`) that:
1. Owns the UART to H2
2. HDLC-encodes outgoing frames and HDLC-decodes incoming frames
3. Routes frames by IID to per-consumer queues
4. Rewrites broadcast IID 3 to each consumer's IID before delivery

### API

```c
// Initialize mux (owns UART)
esp_err_t spinel_mux_init(int uart_port, int tx_pin, int rx_pin, uint32_t baud_rate);

// Register a consumer for an IID (creates per-IID queue)
esp_err_t spinel_mux_register(uint8_t iid, spinel_mux_notify_cb_t notify_cb, void *ctx);

// Pull next frame from IID queue (blocking with timeout)
size_t spinel_mux_recv(uint8_t iid, uint8_t *buf, size_t buf_size, uint32_t timeout_ms);

// Send frame with IID stamped into header
esp_err_t spinel_mux_send(uint8_t iid, const uint8_t *frame, uint16_t frame_len);
```

### RX Architecture

```
UART ISR -> ring buffer -> RX task -> HDLC decode -> extract IID
                                                         |
                                            IID 3 (broadcast)?
                                           /              \
                                         yes              no
                                          |                |
                              for each consumer:     enqueue to
                              rewrite IID 3->N,      IID N queue
                              enqueue to IID N queue
                                          |                |
                                    notify_cb()      notify_cb()
                                   (write eventfd)  (write eventfd)
```

Each consumer has a 4KB FreeRTOS `MessageBuffer` (PSRAM-backed to preserve internal RAM). The notify callback is lightweight — just signals an eventfd so the consumer's select() loop wakes up.

### Broadcast IID Rewrite

This is critical. The H2 NCP sends RESET notifications on IID 3 (broadcast). SpinelDriver on the host only accepts frames matching its configured IID. The mux rewrites the IID in the Spinel header **before enqueuing**:

```c
if (iid == SPINEL_MUX_IID_BROADCAST) {
    for (uint8_t i = 0; i < SPINEL_MUX_IID_BROADCAST; i++) {
        if (s_mux.consumers[i].active) {
            s_mux.rx_frame[0] = SPINEL_SET_IID(header, i);  // rewrite
            enqueue_frame(i, s_mux.rx_frame, payload_len);
        }
    }
}
```

## 5. S3 Host: MuxSpinelInterface

`MuxSpinelInterface` is a C++ class that implements OpenThread's `SpinelInterface` backed by the spinel mux instead of a direct UART.

Both Zigbee (IID 0) and Thread (IID 1) use separate instances of this class. The key methods:

- **`Enable(iid)`**: Registers on the mux with an eventfd-signaling notify callback
- **`SendFrame()`**: Calls `spinel_mux_send(m_iid, frame, len)` — mux stamps the IID
- **`Process()`**: Drains `spinel_mux_recv()` into `RxFrameBuffer` via `WriteByte()` for SpinelDriver consumption
- **`WaitForFrame()`**: Uses `select()` on the eventfd with timeout

### Important: WriteByte, not SaveFrame

SpinelDriver expects to manage the RxFrameBuffer lifecycle itself. Process() must replay frames byte-by-byte via `WriteByte()` and invoke the callback. SpinelDriver then calls `SaveFrame()` or `DiscardFrame()`:

```cpp
void MuxSpinelInterface::Process(const void *aMainloopContext) {
    // ... drain eventfd ...
    while (true) {
        size_t len = spinel_mux_recv(m_iid, s_recv_buf, sizeof(s_recv_buf), 0);
        if (len == 0) break;
        
        bool ok = true;
        for (size_t i = 0; i < len && ok; i++) {
            if (m_frame_buffer->WriteByte(s_recv_buf[i]) != OT_ERROR_NONE)
                ok = false;
        }
        if (ok) m_callback(m_callback_context);
        else m_frame_buffer->DiscardFrame();
    }
}
```

## 6. ESP-IDF Platform Patches Required

Four files in ESP-IDF must be patched for multi-instance support. These are the bugs we found through live debugging on hardware.

### Patch 1: Per-Instance Alarms (ROOT CAUSE of IID 1 TX failure)

**File**: `components/openthread/src/port/esp_openthread_alarm.c`

**Problem**: Stock implementation uses a single global `s_alarm_ms` / `s_is_ms_running`. When instance 1 sets a CSMA backoff timer, instance 0 overwrites it. When the alarm fires, `otPlatAlarmMilliFired(instance0)` is called — instance 1's timer never fires, so `SubMac::BeginTransmit()` is never reached.

**Fix**: Replace single alarm state with per-instance slots:

```c
typedef struct {
    otInstance *instance;
    uint32_t   alarm_ms;
    bool       is_ms_running;
    uint32_t   alarm_us;
    bool       is_us_running;
} alarm_state_t;

static alarm_state_t s_alarms[MAX_ALARM_INSTANCES];
```

Each `otPlatAlarmMilliStartAt(aInstance, ...)` finds/creates a slot for that instance. `esp_openthread_alarm_process()` iterates all slots and fires callbacks with the correct instance.

### Patch 2: TX Completion Routing

**File**: `components/openthread/src/port/esp_openthread_radio.c`

**Problem**: `esp_openthread_radio_process()` calls `otPlatRadioTxDone(aInstance, ...)` with the mainloop instance (always instance 0), even if instance 1 initiated the TX.

**Fix**: Track which instance started the TX:

```c
static otInstance *s_tx_instance = NULL;

otError otPlatRadioTransmit(otInstance *aInstance, otRadioFrame *aFrame) {
    s_tx_instance = aInstance;  // remember who started this TX
    // ... existing TX code ...
}

// In esp_openthread_radio_process():
otInstance *txInst = s_tx_instance ? s_tx_instance : aInstance;
s_tx_instance = NULL;
otPlatRadioTxDone(txInst, &s_transmit_frame, ...);
```

### Patch 3: Instance Storage

**File**: `components/openthread/src/esp_openthread_platform.cpp`

**Problem**: `esp_openthread_get_instance()` calls `ot::Instance::Get()` which doesn't work with multiple static instances.

**Fix**: Store the instance pointer explicitly:

```cpp
static otInstance *s_ot_instance = NULL;

extern "C" void esp_openthread_set_instance(otInstance *instance) {
    s_ot_instance = instance;
}

otInstance *esp_openthread_get_instance(void) {
#if OPENTHREAD_CONFIG_MULTIPLE_INSTANCE_ENABLE
    return s_ot_instance;
#else
    return (otInstance *)&ot::Instance::Get();
#endif
}
```

### Patch 4: Multi-Instance NCP Init

**File**: `components/openthread/src/ncp/esp_openthread_ncp_hdlc.cpp`

**Problem**: IDF's NCP HDLC wrapper only exposes single-instance `otNcpHdlcInit()`. The upstream OpenThread has `otNcpHdlcInitMulti()` but IDF's override file doesn't include it.

**Fix**: Add the multi-instance version:

```cpp
void otNcpHdlcInitMulti(otInstance **aInstances, uint8_t aCount,
                         otNcpHdlcSendCallback aSendCallback) {
    // Initialize NCP with multiple instances
}
```

### Build System Patch (for Combo Mode)

**File**: `components/openthread/CMakeLists.txt`

When both OpenThread FTD and Zigbee spinel are enabled, the Zigbee radio spinel sources (`src/spinel/`) must be added to the FTD build:

```cmake
if(CONFIG_OPENTHREAD_ENABLED AND CONFIG_THERMOSTAT_ENABLE_SPINEL_MUX)
    list(APPEND src_dirs "src/spinel")
endif()
```

Also enable vendor hooks in the FTD config (`openthread-core-esp32x-ftd-config.h`):
```c
#if CONFIG_THERMOSTAT_ENABLE_SPINEL_MUX
#define OPENTHREAD_SPINEL_CONFIG_VENDOR_HOOK_ENABLE 1
#endif
```

And guard the duplicate `otLinkGetFrameCounter` in `esp_radio_spinel.cpp`:
```c
#if !OPENTHREAD_FTD && !OPENTHREAD_MTD
uint32_t otLinkGetFrameCounter(otInstance *aInstance) { ... }
#endif
```

## 7. Build Configuration

### Kconfig Options

```
choice THERMOSTAT_RADIO_MODE
    config THERMOSTAT_RADIO_ZIGBEE_ONLY
        bool "Zigbee coordinator only"
    config THERMOSTAT_RADIO_THREAD_BR_ONLY
        bool "Thread Border Router only"
    config THERMOSTAT_RADIO_COMBO
        bool "Zigbee + Thread (multipan mux)"
endchoice

config THERMOSTAT_ENABLE_SPINEL_MUX
    bool
    default y if THERMOSTAT_RADIO_COMBO || THERMOSTAT_RADIO_ZIGBEE_ONLY || THERMOSTAT_RADIO_THREAD_BR_ONLY
```

### Combo sdkconfig Overlay

```ini
CONFIG_THERMOSTAT_RADIO_COMBO=y
CONFIG_ZB_ENABLED=y
CONFIG_ZB_ZCZR=y
CONFIG_ZB_RADIO_SPINEL=y
CONFIG_ZB_RADIO_SPINEL_UART=y
CONFIG_OPENTHREAD_ENABLED=y
CONFIG_OPENTHREAD_FTD=y
CONFIG_OPENTHREAD_BORDER_ROUTER=y
CONFIG_OPENTHREAD_RADIO_SPINEL_UART=y
CONFIG_THERMOSTAT_ENABLE_SPINEL_MUX=y
```

## 8. Debugging Guide

### Enable H2 Trace Logging

Add to H2 sdkconfig: `CONFIG_THERMOSTAT_H2_MULTIPAN_TRACE=y`

This enables NCP TX frame logging and mainloop iteration logging.

### Key Log Tags

| Tag | Source | What It Shows |
|-----|--------|---------------|
| `spinel_mux` | S3 mux | Frame routing, drops, CRC errors |
| `mux_spinel` | MuxSpinelInterface | Enable/disable, frame delivery |
| `thread_br` | Thread BR init | Heap state, WiFi wait, OT start |
| `zigbee_gw` | Zigbee gateway | Network formation, device joins |
| `OPENTHREAD` | OT stack | Role changes, TX timeouts |
| `NCP_RAW` | H2 NCP (trace) | STREAM_RAW handler, TX_DONE |

### The TX Timeout Debug Method

If Thread shows `radio tx timeout`:

1. Add `ESP_LOGI` to `HandlePropertySet_SPINEL_PROP_STREAM_RAW` on H2 — check if STREAM_RAW is received and `rawEnabled=1`
2. Add `ESP_LOGI` to `otPlatRadioTransmit` on H2 — check if the platform radio TX is called
3. If STREAM_RAW enters but `otPlatRadioTransmit` never fires → alarm issue (CSMA timer not firing)
4. If `otPlatRadioTransmit` fires but no `TX_DONE` → radio completion routing issue
5. Check `otPlatAlarmMilliFired` is called with the correct instance

## 9. Known Limitations

1. **Single channel**: Both Zigbee and Thread must operate on the same 802.15.4 channel (no concurrent listening on ESP32-H2)
2. **Internal RAM pressure**: In combo mode, WiFi/LWIP/mDNS consume most internal SRAM. OT task needs internal stack for AES crypto (PSRAM crashes DMA). Careful heap management required.
3. **No hot-swap**: Both protocol stacks must be initialized at boot. Dynamic enable/disable not implemented.
4. **Silicon Labs abandoned this**: Their multipan was marked "experimental" and community recommends separate radios. Our implementation is more ambitious.

## 10. How We Got Here: The Investigation

### Timeline

1. **Zigbee-only on mux**: Worked immediately on IID 0 (default active radio instance)
2. **Thread-only on mux**: Property queries worked on IID 1, but **radio TX timed out**
3. **Root cause hunt**: Added ESP_LOGI instrumentation across 4 layers:
   - NCP handler: STREAM_RAW received, `rawEnabled=1`, `otLinkRawTransmit` returns success
   - Platform radio: `otPlatRadioTransmit` **never called** despite TX initiated
   - SubMac trace: `Send()` → `StartCsmaBackoff()` → `StartTimerForBackoff()` — timer-based, not direct call
   - Alarm system: **Single global alarm** — instance 1's timer overwritten by instance 0
4. **Fix 1**: Rewrote `esp_openthread_alarm.c` with per-instance slots → `otPlatRadioTransmit` now called
5. **Fix 2**: Added `s_tx_instance` tracking → `otPlatRadioTxDone` routes to correct instance
6. **Fix 3**: Mainloop processes all instances' tasklets
7. **Verification**: H2 console shows `STREAM_RAW → otLinkRawTransmit → TX_DONE: iid=1 error=0`
8. **Mux refactor**: Replaced shared-buffer callback with per-IID MessageBuffer queues
9. **Combo build**: Added CMake/Kconfig plumbing for FTD + Zigbee spinel coexistence

### What Other Platforms Get Wrong

We researched Silicon Labs, Nordic, TI, and the OpenThread reference implementation:

- **Every platform** uses a single global alarm in their `otPlatAlarmMilliStartAt` implementation
- **Every platform** ignores the `aInstance` parameter (marked `OT_UNUSED_VARIABLE`)
- **No platform** documents the per-instance requirements for multipan
- **The OpenThread API docs** don't specify that `otPlatAlarmMilliFired` must be called with the same instance that called `StartAt`

This is a systemic gap in the OpenThread platform abstraction documentation.

---

## Contributing

This guide is based on a working implementation running on:
- **Host**: ESP32-S3 (Waveshare ESP32-S3-Touch-LCD-4.3B)
- **RCP**: ESP32-H2 (DevKit)
- **IDF**: ESP-IDF v6.1-dev
- **OpenThread**: Bundled with IDF 6.1

The platform patches should be submitted as PRs to [espressif/esp-idf](https://github.com/espressif/esp-idf). The most impactful contribution is the per-instance alarm fix (`esp_openthread_alarm.c`) — without it, multipan TX on any non-default IID is fundamentally broken.

Relevant upstream issues:
- [espressif/esp-idf#13815](https://github.com/espressif/esp-idf/issues/13815) — "Integrating Thread RCP and Zigbee RCP functions in ESP32-H2"
- [openthread/openthread#9360](https://github.com/openthread/openthread/pull/9360) — Multiple spinel instances support
